pkg/ebpf/http.go

```diff
@@ -0,0 +1,196 @@
+// GET /foo HTTP/1.1
+func (tx *httpTX) Path() string {
+	b := C.GoBytes(unsafe.Pointer(&tx.request_fragment), C.int(C.HTTP_BUFFER_SIZE))
+
+	var i, j int
+	for i = 0; i < len(b) && b[i] != ' '; i++ {
+	}
+
+	i++
+
+	for j = i; j < len(b) && b[j] != ' '; j++ {
+	}
+
+	if i < j && j <= len(b) {
+		return string(b[i:j])
+	}
+
+	return ""
+}
+
+// IsDirty detects whether the batch page we're supposed to read from is still
+// valid.  A "dirty" page here means that between the time the
+// http_notification_t message was sent to userspace and the time we performed
+// the batch lookup the page was overriden.
+func (batch *httpBatch) IsDirty(notif httpNotification) bool {
+	return int(batch.idx) >= int(notif.batch_idx)+HTTPBatchPages
+}
+
+// GetTransactions extracts the HTTP transactions from the batch acording to the
+// httpNotification received from the Kernel.
+func (batch *httpBatch) GetTransactions(notif httpNotification) *[HTTPBatchSize]httpTX {
+	pageID := int(notif.batch_idx) % HTTPBatchPages
+	return (*[HTTPBatchSize]httpTX)(unsafe.Pointer(&batch.txs[pageID*HTTPBatchSize]))
+}
+
+type httpMonitor struct {
+	batchMap      *ebpf.Map
+	perfMap       *manager.PerfMap
+	perfHandler   *bytecode.PerfHandler
+	closeFilterFn func()
+}
+
+func newHTTPMonitor(config *Config, m *manager.Manager, h *bytecode.PerfHandler) (*httpMonitor, error) {
+	filter, _ := m.GetProbe(manager.ProbeIdentificationPair{Section: string(bytecode.SocketHTTPFilter)})
+	if filter == nil {
+		return nil, fmt.Errorf("error retrieving socket filter")
+	}
+
+	closeFilterFn, err := network.HeadlessSocketFilter(config.ProcRoot, filter)
+	if err != nil {
+		return nil, fmt.Errorf("error enabling HTTP traffic inspection: %s", err)
+	}
+
+	batchMap, _, err := m.GetMap(string(bytecode.HttpBatchMap))
+	if err != nil {
+		return nil, err
+	}
+
+	notificationMap, _, _ := m.GetMap(string(bytecode.HttpEventMap))
+	numCPUs := int(notificationMap.ABI().MaxEntries)
+	for i := 0; i < numCPUs; i++ {
+		batch := new(httpBatch)
+		batchMap.Put(unsafe.Pointer(&i), unsafe.Pointer(batch))
+	}
+
+	pm, found := m.GetPerfMap(string(bytecode.HttpEventMap))
+	if !found {
+		return nil, fmt.Errorf("unable to find perf map %s", bytecode.HttpEventMap)
+	}
+
+	return &httpMonitor{
+		batchMap:      batchMap,
+		perfMap:       pm,
+		perfHandler:   h,
+		closeFilterFn: closeFilterFn,
+	}, nil
+}
+
+// Start consuming HTTP events
+// Please note the code below is merely an *example* of how to consume the HTTP
+// transaction information sent from the eBPF program
+func (http *httpMonitor) Start() error {
+	if http == nil {
+		return nil
+	}
+
+	if err := http.perfMap.Start(); err != nil {
+		return fmt.Errorf("error starting perf map: %s", err)
+	}
+
+	go func() {
+		var (
+			misses int
+			then   = time.Now()
+			hits   = make(map[int]int)
+			report = time.NewTicker(30 * time.Second)
+		)
+
+		for {
+			select {
+			case data, ok := <-http.perfHandler.ClosedChannel:
```

Comment:
We probably want to rename `ClosedChannel` to something more generic, since Closed doesn't make much sense in this context.